In [126]:
# Import Dependencies
import pandas as pd
import numpy as np

In [127]:
# load in files

lab_file = pd.read_csv("Resources_raw/lab_data_raw.csv")
rx_file = pd.read_csv("Resources_raw/reactor_data_raw.csv")
ta_file = pd.read_csv("Resources_raw/towerA_data_raw.csv")
tb_file = pd.read_csv("Resources_raw/towerB_data_raw.csv")
tc_file = pd.read_csv("Resources_raw/towerC_data_raw.csv")

# csv to dataframe

lab_data = pd.DataFrame(lab_file)
rx_data = pd.DataFrame(rx_file)
ta_data = pd.DataFrame(ta_file)
tb_data = pd.DataFrame(tb_file)
tc_data = pd.DataFrame(tc_file)

In [128]:
# check dataframes loaded in correctly
lab_data.head()

Date TA.OVH.RECYCLE.NC4.VOLPCT
0  7/15/22 13:00               11.19999981
1   7/15/22 6:00               13.89999962
2  7/14/22 13:00               13.60000038
3   7/14/22 6:00                        11
4  7/13/22 13:00               10.69999981

In [129]:
# count rows
print(len(lab_data.index))

15589


In [130]:
rx_data.head()

Date PREFRAC RECYCLE TO CONT PREFRAC RECYCLE TO CONT.1  \
0  7/15/2022 13:00              2129.64624                        18   
1   7/15/2022 6:00                    2272                        18   
2  7/14/2022 13:00             2201.021484                        18   
3   7/14/2022 6:00             2298.876953                        18   
4  7/13/2022 13:00             2269.168213                        18   

  ALKY NORTH CONTACT. FEED ALKY NORTH CONTACT. FEED.1 NORTH CONTACTOR FLUSH  \
0              2675.764893                 18.6390686            154.651001   
1              2525.604736                16.98755646           154.9970703   
2              2205.657471                13.89487362            154.587738   
3               2419.37085                15.55793953           153.0914307   
4              2569.280518                17.27506256            149.634552   

  NORTH RX OUTLET TEMP DEISO RECYCLE TO CONTACT DEISO RECYCLE TO CONTACT.1  \
0          98.29890442              6399.555664                51.66233826   
1          94.09553528              6395.504883                51.38218689   
2          97.24479675              5933.965332                49.18594742   
3          94.88616943              5902.112793                49.26596451   
4          99.74843597               5749.19873                48.75584412   

  ALKY SOUTH CONTACT. FEED ALKY SOUTH CONTACT. FEED.1 SOUTH CONTACTOR FLUSH  \
0               2656.32959                18.04207611            143.763855   
1              2514.057617                16.42683601           145.5339355   
2              2414.575195                          0           145.1560974   
3              2375.452881                14.67038441           143.6205597   
4              2567.080566                 16.4713192           141.7488098   

  SOUTH RX OUTLET TEMP SOUTH CONTACT C.W.R. CONDENSATE PH MONITOR  \
0          101.0686264          83.57925415             6.6152215   
1          97.15179443          80.87692261           6.616475582   
2          101.5997391           85.2561264           6.553259373   
3          97.67050171          82.32585144           6.557726383   
4          102.5134277          86.02752686            6.56847477   

  CONTACTOR DELTA T CONTRL  
0             -2.765020609  
1             -3.041529655  
2             -4.411693096  
3             -2.753258467  
4             -2.850717783

In [131]:
print(len(rx_data.index))

22244


In [132]:
ta_data.head()

Date TA.Charge.BPD   TA.OH.PSIG   TA.Tray4.F      TA.OH.F  \
0  7/15/2022 13:00    10731.7002  99.30152893  134.6903534  133.1420898   
1   7/15/2022 6:00    10332.8252  99.47181702  134.7993469  133.2794189   
2  7/14/2022 13:00   9904.880859  98.97205353  134.3727112  132.7810669   
3   7/14/2022 6:00   9641.806641  99.94731903  134.9676514  133.2887726   
4  7/13/2022 13:00   9876.027344  99.60662842  135.2410889  133.3068848   

   TA.RECV.PCT TA.RECV.PSIG TA.REFLUX.BPD TA.RECYCLE.TO.CONTACT.BPD  \
0  1.432772994  100.6223145   5919.916992               6399.555664   
1  3.479724407  99.48440552   5935.238281               6395.504883   
2  2.860878468  99.41940308   5968.467773               5933.965332   
3  4.899298191  99.85849762   6020.976074               5902.112793   
4  1.202635884  100.4339676   5917.750488                5749.19873   

  TA.RECYCLE.TO.CONTACT.F  TA.TRAY39.F  TA.TRAY45.F   TA.DP.PSIG  TA.BTM.PSIG  \
0             97.03623199  160.8286591  177.0243378   0.66986084  98.63166809   
1             93.11539459  159.8332214   175.662384  0.578268826  98.89743805   
2             97.99095917   161.156311    178.83255  0.781860352  98.19156647   
3              92.6629715  159.7163849  176.0884705  0.605773926  99.33746338   
4             98.54929352  163.9949493  182.7541046   0.64453125  98.95614624   

  TA.BTM.LEVEL.PCT TA.BTM.REBOIL.OUT.F TA.TOP.REBOIL.OUT.F  
0      74.77849579         315.3513489         188.9165344  
1      72.92869568         312.3144531         188.0589294  
2      73.68263245         318.3023071         191.8672333  
3      74.87386322         312.1195374         187.9542999  
4       76.2322998         318.6087952         195.6160431

In [133]:
print(len(ta_data.index))

15589


In [134]:
tb_data.head()

Date TB.CHARGE.BPD TB.CHARGE.PREHEAT.COND.LBHR  TB.CHARGE.F  \
0  7/15/2022 13:00   11822.55273                 3364.362061  153.4123688   
1   7/15/2022 6:00   12301.79785                 3932.324463  153.3493042   
2  7/14/2022 13:00   11201.09961                 3097.979004  153.1002808   
3   7/14/2022 6:00   11154.17871                 3472.200684  152.6564484   
4  7/13/2022 13:00   10675.06152                 2829.445557  154.0720673   

     TB.OVHD.F TB.OVHD.PSIG TB.OVHD.REVEIVER.PSIG TB.OVHD.REVEIVER.F  \
0  175.7751617  183.1718445           182.9406891        98.85330963   
1  179.4546509  183.0008698           182.9354095        77.24609375   
2  178.3122253  183.1327057           182.8491211        100.2281265   
3  181.3551788  183.0615234           183.4066772        86.05672455   
4  176.9775391  183.8012848           183.6343384        100.1552429   

  TB.OVHD.REVEIVER.LVL.PCT  TB.BTM.PSIG TB.BTM.LVL.PCT    TB.DP.PSIG  \
0              56.35223389  184.3353271    80.71112061   1.184082031   
1              19.95126724  187.2182465    79.10299683   4.237060547   
2              52.53837967  183.0235291    79.96920776  -0.141448975   
3              58.93585587  183.4649658    79.71007538   0.408325195   
4              29.10192108  187.3635864    79.88949585   3.549194336   

      TB.BTM.F TB.REBOIL.OUT.F  
0  213.6970673     216.1233521  
1  216.1178894     218.4076538  
2  214.6928711     217.1874542  
3  215.4292297     217.7776794  
4  218.8019867     221.3375092

In [135]:
print(len(tb_data.index))

22248


In [136]:
tc_data.tail()

Date TC.TRAY7.F TC.TRAY40.F TC.CHARGE.F TC.OVHD.PSIG TC.OVHD.RECVR.F  \
22243  NaN        NaN         NaN         NaN          NaN             NaN   
22244  NaN        NaN         NaN         NaN          NaN             NaN   
22245  NaN        NaN         NaN         NaN          NaN             NaN   
22246  NaN        NaN         NaN         NaN          NaN             NaN   
22247  NaN        NaN         NaN         NaN          NaN             NaN   

      TC.TRAY24.F TC.DP.PSIG TC.BTM.PSIG TC.REBOIL.OUT.F TC.REBOIL.COND.LBHR  
22243         NaN        NaN         NaN             NaN                 NaN  
22244         NaN        NaN         NaN             NaN                 NaN  
22245         NaN        NaN         NaN             NaN                 NaN  
22246         NaN        NaN         NaN             NaN                 NaN  
22247         NaN        NaN         NaN             NaN                 NaN

In [137]:
print(len(tc_data.index))

22248


### Notes
- Towers C and B as well as the Reactors have 22,248 rows
- Tower A and the Lab data have 15,589 rows
- Compare dates to check if more data was pulled for C, B, and RX data
- Also, Turnaround for the plant occurred between 10/12/2018 and 3/29/2019 so all rows between these dates need to be removed from each dataset
- Checkpoints are taken at hour 6:00 and hour 13:00, all other times will be considered invalid and will be removed

In [138]:
# tc_data.tail() revealed a ton of null values, will drop them

In [139]:
# preparing list of feature names for use in replacing text strings
# with nulls (will later drop these nulls and mass convert the dataframe objects)
# to floats
lab_columns = list(lab_data)
rx_columns = list(rx_data)
ta_columns = list(ta_data)
tb_columns = list(tb_data)
tc_columns = list(tc_data)

#print(lab_columns)
#print(rx_columns)
#print(ta_columns)
#print(tb_columns)
#print(tc_columns)

### Tentative Column Name Changes
- column names will need to be changed for better understanding

#### Lab Columns
- Date will remain unchanged
- TA.OVH.RECYCLE.NC4.VOLPCT = Butane

#### Reactor Columns
- Date will remain unchanged
- REACTOR Features transformation will be put on hold until I understand this better

#### Tower A Columns
- Date will remain unchanged
- TA.Charge.BPD = Charge_bpd
- TA.OH.PSIG = Overhead_psig
- TA.Tray4.F = Tray4_temp
- TA.OH.F = Overhead_temp
- TA.RECV.PCT = Received_pct
- TA.RECV.PSIG = Received_psig
- TA.REFLUX.BPD = Reflux_bpd
- TA.RECYCLE.TO.CONTACT.BPD = Recycle_bpd
- TA.RECYCLE.TO.CONTACT.F = Recycle_temp
- TA.TRAY39.F = Tray39_temp
- TA.TRAY45.F = Tray45_temp
- TA.DP.PSIG = Differential_psig
- TA.BTM.PSIG = Bottom_psig
- TA.BTM.LEVEL.PCT = Btm_lvl_pct
- TA.BTM.REBOIL.OUT.F = Btm_reboil_temp
- TA.TOP.REBOIL.OUT.F = top_reboil_temp

#### Tower B Columns
- Date will remain unchanged
- TB.CHARGE.BPD = Charge_bpd
- TB.CHARGE.PREHEAT.COND.LBHR = "Preheat Condensor Flow Rate LB/HR" (*)
- TB.CHARGE.F = Charge_temp
- TB.OVHD.F = Overhead_temp
- TB.OVHD.PSIG = Overhead_psig
- TB.OVHD.REVEIVER.PSIG = OH_receiver_psig
- TB.OVHD.REVEIVER.F = OH_reciever_temp
- TB.OVHD.REVEIVER.LVL.PCT = OH_receiver_pct
- TB.BTM.PSIG = Bottom_psig
- TB.BTM.LVL.PCT = Btm_lvl_pct
- TB.DP.PSIG = Differential_psig
- TB.BTM.F = Bottom_temp
- TB.REBOIL.OUT.F = Reboil_temp

#### Tower C Columns
- Date will remain unchanged
- TC.TRAY7.F = Tray7_temp
- TC.TRAY40.F = Tray40_temp
- TC.CHARGE.F = Charge_temp
- TC.OVHD.PSIG = Overhead_psig
- TC.OVHD.RECVR.F = OH_receiver_temp
- TC.TRAY24.F = Tray24_temp
- TC.DP.PSIG = Differential_psig
- TC.BTM.PSIG = Bottom_psig
- TC.REBOIL.OUT.F = Reboil_temp
- TC.REBOIL.COND.LBHR = "Reboiler Cooling Water Flow Rate"

# Clean Lab Data

In [140]:
# convert Date column from object to datetime
lab_data['Date'] = pd.to_datetime(lab_data['Date'])
lab_data.head(1)

Date TA.OVH.RECYCLE.NC4.VOLPCT
0 2022-07-15 13:00:00               11.19999981

In [141]:
# add hour column to sort out invalid checkpoint hours
lab_data['Hour'] = pd.to_datetime(lab_data['Date']).dt.hour
lab_data.head(1)

Date TA.OVH.RECYCLE.NC4.VOLPCT  Hour
0 2022-07-15 13:00:00               11.19999981    13

In [142]:
# keep only rows that have valid hour checkpoints
lab_data = lab_data.loc[(lab_data['Hour'] == 6) | (lab_data['Hour'] == 13)]
lab_data

Date TA.OVH.RECYCLE.NC4.VOLPCT  Hour
0     2022-07-15 13:00:00               11.19999981    13
1     2022-07-15 06:00:00               13.89999962     6
2     2022-07-14 13:00:00               13.60000038    13
3     2022-07-14 06:00:00                        11     6
4     2022-07-13 13:00:00               10.69999981    13
...                   ...                       ...   ...
15584 2013-01-06 06:00:00                        18     6
15585 2013-01-05 13:00:00               14.39999962    13
15586 2013-01-05 06:00:00               10.19999981     6
15587 2013-01-04 13:00:00               6.099999905    13
15588 2013-01-04 06:00:00               4.900000095     6

[6547 rows x 3 columns]

In [143]:
# store start and end dates of turnaround as variables
start = "2018-10-12 13:00:00"
end = "2019-03-29 6:00:00"

# turn datetimes between start and end into nulls
lab_data.loc[(lab_data['Date'] > start) & (lab_data['Date'] <= end)] = np.nan

In [144]:
# for loop to nullify text strings
for name in lab_columns:
    lab_data.loc[lab_data[name] == "I/O Timeout"] = np.nan
    lab_data.loc[lab_data[name] == "Error"] = np.nan
    lab_data.loc[lab_data[name] == "Bad"] = np.nan
    lab_data.loc[lab_data[name] == "Bad Input"] = np.nan

In [145]:
# check nulls
lab_data.isnull().sum()

Date                         1205
TA.OVH.RECYCLE.NC4.VOLPCT    1205
Hour                         1205
dtype: int64

In [146]:
# drop nulls from dataframe
lab_data = lab_data.dropna()
lab_data

Date TA.OVH.RECYCLE.NC4.VOLPCT  Hour
0     2022-07-15 13:00:00               11.19999981  13.0
1     2022-07-15 06:00:00               13.89999962   6.0
2     2022-07-14 13:00:00               13.60000038  13.0
3     2022-07-14 06:00:00                        11   6.0
4     2022-07-13 13:00:00               10.69999981  13.0
...                   ...                       ...   ...
15584 2013-01-06 06:00:00                        18   6.0
15585 2013-01-05 13:00:00               14.39999962  13.0
15586 2013-01-05 06:00:00               10.19999981   6.0
15587 2013-01-04 13:00:00               6.099999905  13.0
15588 2013-01-04 06:00:00               4.900000095   6.0

[5342 rows x 3 columns]

In [147]:
# check data types
lab_data.dtypes

Date                         datetime64[ns]
TA.OVH.RECYCLE.NC4.VOLPCT            object
Hour                                float64
dtype: object

In [148]:
# convert object types to float types
lab_data['TA.OVH.RECYCLE.NC4.VOLPCT'] = lab_data['TA.OVH.RECYCLE.NC4.VOLPCT'].astype(float)
lab_data.dtypes

C:\Users\anna\AppData\Local\Temp/ipykernel_2600/251440205.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  lab_data['TA.OVH.RECYCLE.NC4.VOLPCT'] = lab_data['TA.OVH.RECYCLE.NC4.VOLPCT'].astype(float)


Date                         datetime64[ns]
TA.OVH.RECYCLE.NC4.VOLPCT           float64
Hour                                float64
dtype: object

In [149]:
# reset index 
lab_data = lab_data.reset_index()
#lab_data

In [150]:
# drop hour column
#lab_data = lab_data.drop(columns=['Hour', 'index'])
lab_data.head()

index                Date  TA.OVH.RECYCLE.NC4.VOLPCT  Hour
0      0 2022-07-15 13:00:00                       11.2  13.0
1      1 2022-07-15 06:00:00                       13.9   6.0
2      2 2022-07-14 13:00:00                       13.6  13.0
3      3 2022-07-14 06:00:00                       11.0   6.0
4      4 2022-07-13 13:00:00                       10.7  13.0

In [151]:
# summary statistics
lab_data.describe()

index  TA.OVH.RECYCLE.NC4.VOLPCT         Hour
count   5342.000000                5342.000000  5342.000000
mean    9353.236054                  11.136153     9.484276
std     6183.562042                  14.474349     3.500292
min        0.000000                   0.000000     6.000000
25%     1357.250000                   8.200000     6.000000
50%    12866.500000                  10.300000     6.000000
75%    14223.750000                  12.800000    13.000000
max    15588.000000                1005.000000    13.000000

In [152]:
# define find outlier function
def find_outliers(df):
    q1 = df.quantile(0.25)
    q3 = df.quantile(.75)
    iqr = q3-q1
    lower_lim = q1 - (1.5 * iqr)
    upper_lim = q3 + (1.5 * iqr)
    outliers = df[((df<lower_lim) | (df>upper_lim))]
    return outliers

In [153]:
# run function on lab data
lab_outliers = find_outliers(lab_data['TA.OVH.RECYCLE.NC4.VOLPCT'])
lab_outliers.describe()

count     224.000000
mean       24.769196
std        67.705369
min         0.000000
25%        19.799999
50%        21.799999
75%        26.125000
max      1005.000000
Name: TA.OVH.RECYCLE.NC4.VOLPCT, dtype: float64

Through discussion, we have determined that values above 40 will be considered invalid and should be removed. All other values are within reasonable measure, regardless of statistical outliers.

In [154]:
# find invalid values and nullify
lab_data.loc[lab_data['TA.OVH.RECYCLE.NC4.VOLPCT'] >= 40] = np.nan

In [155]:
# drop the nulls
lab_data = lab_data.dropna()

In [156]:
# rerun outlier function
lab_outliers = find_outliers(lab_data['TA.OVH.RECYCLE.NC4.VOLPCT'])
lab_outliers.describe()

count    211.000000
mean      18.107109
std       11.065566
min        0.000000
25%        1.100000
50%       21.600000
75%       25.250000
max       36.799999
Name: TA.OVH.RECYCLE.NC4.VOLPCT, dtype: float64

In [158]:
# export cleaned data into a new csv
#lab_data.to_csv("Resources_clean/lab_data.csv")